nifi-nar-bundles/nifi-stateful-analysis-bundle/nifi-stateful-analysis-processors/src/main/java/org/apache/nifi/processors/stateful/analysis/AttributeRollingWindow.java

```diff
@@ -0,0 +1,325 @@
+            StateMap state = stateManager.getState(SCOPE);
+            HashMap<String, String> tempMap = new HashMap<>();
+            tempMap.putAll(state.toMap());
+            if (!tempMap.containsKey(COUNT_KEY)) {
+                tempMap.put(COUNT_KEY, "0");
+                context.getStateManager().setState(tempMap, SCOPE);
+            }
+        }
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        try {
+            Long currTime = System.currentTimeMillis();
+            if(microBatchTime == null){
+                noMicroBatch(context, session, flowFile, currTime);
+            } else{
+                microBatch(context, session, flowFile, currTime);
+            }
+
+        } catch (IOException e) {
+            getLogger().error("Failed to save the state for {} due to {}", new Object[] { flowFile, e});
+            session.transfer(flowFile, REL_FAILURE);
+        }
+    }
+
+    private void noMicroBatch(ProcessContext context, ProcessSession session, FlowFile flowFile, Long currTime) throws IOException {
+        final StateManager stateManager = context.getStateManager();
+
+        Map<String, String> state = new HashMap<>(stateManager.getState(SCOPE).toMap());
+        Long count = Long.valueOf(state.get(COUNT_KEY));
+        count ++;
+
+        Set<String> keysToRemove = new HashSet<>();
+
+        for(String key: state.keySet()){
+            if(!key.equals(COUNT_KEY)){
+                Long timeStamp = Long.decode(key);
+
+                if(currTime - timeStamp > timeWindow) {
+                    keysToRemove.add(key);
+                    count --;
+
+                }
+            }
+        }
+        String countString = String.valueOf(count);
+
+        for(String key: keysToRemove){
+            state.remove(key);
+        }
+
+        Double aggregateValue = 0.0D;
+
+        for(Map.Entry<String,String> entry: state.entrySet()){
+            if(!entry.getKey().equals(COUNT_KEY)){
+                aggregateValue += Double.valueOf(entry.getValue());
+            }
+        }
+
+
+        final Double currentFlowFileValue = context.getProperty(VALUE_TO_TRACK).evaluateAttributeExpressions(flowFile).asDouble();
+
+        aggregateValue += currentFlowFileValue;
+
+        state.put(String.valueOf(currTime), String.valueOf(currentFlowFileValue));
+
+        state.put(COUNT_KEY, countString);
+        stateManager.setState(state, SCOPE);
+
+        Map<String, String> attributesToAdd = new HashMap<>();
+        attributesToAdd.put(ROLLING_WINDOW_VALUE_KEY, String.valueOf(aggregateValue));
+        attributesToAdd.put(ROLLING_WINDOW_COUNT_KEY, String.valueOf(count));
+
+        flowFile = session.putAllAttributes(flowFile, attributesToAdd);
+
+        session.transfer(flowFile, REL_SUCCESS);
```

Comment:
In the absence of any other provenance event being emitted and attributes are modified the session will automatically emit a modifyAttribute provenance event.